lib/setopt.c

```diff
@@ -631,6 +631,20 @@ CURLcode Curl_vsetopt(struct Curl_easy *data, CURLoption option, va_list param)
     data->set.method = HTTPREQ_POST_FORM;
     data->set.opt_no_body = FALSE; /* this is implied */
     break;
+
+  case CURLOPT_V4_SIGNATURE:
+    /*
+     * String that holds file type of the SSL certificate to use
+     */
+    result = Curl_setstropt(&data->set.str[STRING_V4_SIGNATURE],
+                            va_arg(param, char *));
+    /*
+     * Basic been set by default it need to be unset here
+     */
+    if(data->set.str[STRING_V4_SIGNATURE])
+      data->set.httpauth = CURLAUTH_SIGNATURE_V4;
```

Comment:
Why should this set this auth bit uniquely in httpauth? Can't user opt to use more than one? Why not use the normal httpauth way?

Reply:
I don't know,
It seems ```pickoneauth``` in http.c choose the first authmode it see, so if I use 
```data->set.httpauth |= CURLAUTH_SIGNATURE_V4``` httpauth contain both CURLAUTH_SIGNATURE_V4 and CURLAUTH_BASIC, and then pickoneauth return the current authmode as basic, which ignore v4_signature.
Also, I modify data->set.httpauth here, because if CURLAUTH_SIGNATURE_V4  is call, I guess it's safe to assume the user want v4 signature as authmode. and it keep the user from calling both CURLOPT_HTTPAUTH and CURLOPT_V4_SIGNATURE.
(though by saying that, I've just saw that I've forgot to update CURLOPT_HTTPAUTH doc)

Reply:
> I still have some questions around how this fits into `CURLOPT_HTTPAUTH`. Your code now sets the v4sig bit _exclusively_ if the v4sig option is set, which seems to imply that this auth is more important than the others or doesn't play on the same level. HTTP authorization is typically the client using one out of possibly a set of the different ones that the server offers.

The problem is that it seems that AWS return an error 400 instead of a 401 when Authorization isn't set properly, and so we can't guess which auth to use from the error.
So sadly we have to force CURLOPT_HTTPAUTH if data->set.httpauth isn't set.